fifteen_min/src/viewer.rs

```diff
@@ -277,3 +224,61 @@ fn build_panel(
         .aligned(HorizontalAlignment::Right, VerticalAlignment::Top)
         .build(ctx)
 }
+
+struct HoverOnBuilding {
+    tooltip: Text,
+    drawn_route: Drawable,
+}
+/// (building, scale factor)
+type HoverKey = (BuildingID, f64);
+
+impl HoverOnBuilding {
+    fn key(ctx: &EventCtx, app: &SimpleApp) -> Option<HoverKey> {
+        match app.mouseover_unzoomed_buildings(ctx) {
+            Some(ID::Building(b)) => {
+                let scale_factor = if ctx.canvas.cam_zoom >= app.opts.min_zoom_for_detail {
+                    1.0
+                } else {
+                    10.0
+                };
+                Some((b, scale_factor))
+            }
+            _ => None,
+        }
+    }
+
+    fn value(
+        ctx: &mut EventCtx,
+        app: &mut SimpleApp,
+        key: HoverKey,
+        isochrone: &Isochrone,
+    ) -> HoverOnBuilding {
+        debug!("Calculating route for {:?}", key);
+
+        let (hover_id, scale_factor) = key;
+        let mut batch = GeomBatch::new();
+        if let Some(polyline) = isochrone
+            .path_to(&app.map, hover_id)
+            .and_then(|path| path.trace(&app.map, Distance::ZERO, None))
+        {
+            let dashed_lines = polyline.dashed_lines(
+                Distance::meters(0.75 * scale_factor),
+                Distance::meters(1.0 * scale_factor),
+                Distance::meters(0.4 * scale_factor),
+            );
+            batch.extend(Color::BLACK, dashed_lines);
+        }
+
+        // Also update this to conveniently get an outline drawn
+        app.current_selection = Some(ID::Building(hover_id));
```

Comment:
Having `app.current_selection` mutated as a side effect of this method is a little sneaky IMO. I think it was less surprising near the `ctx.redo_mouseover()` call. 

Maybe you agree or maybe not, but if you do, now you can hoist it to the callsite, which is a nice testament to this new API vs #405 👍: 

```
.update(HoverOnBuilding::key(ctx, app), |key| {
    app.current_selection = Some(ID::Building(hover_id));
    HoverOnBuilding::new(ctx, app, key, isochrone)
});
```

Reply:
Ah, good point, thanks! I successfully moved it into the closure, but then realized there's a bug -- we want to set current_selection to None, and the callback only happens for defined keys. So I just moved the setting out of the `update` call entirely.

Reply:
oh right, good catch!